tests/mapper/test_link_mappers.py

```diff
@@ -783,3 +783,390 @@ def test_Attri2VecLinkGenerator_unsupervisedSampler_sample_generation(self):
 
         with pytest.raises(IndexError):
             nf, nl = mapper[8]
+
+
+class Test_DirectedGraphSAGELinkGenerator:
+    """
+    Tests of GraphSAGELinkGenerator class
+    """
+
+    n_feat = 4
+    batch_size = 2
+    in_samples = [2, 4]
+    out_samples = [6, 8]
+
+    def test_LinkMapper_constructor(self):
+
+        G = example_graph(feature_size=self.n_feat, is_directed=True)
+        edge_labels = [0] * G.number_of_edges()
+
+        generator = DirectedGraphSAGELinkGenerator(
+            G,
+            batch_size=self.batch_size,
+            in_samples=self.in_samples,
+            out_samples=self.out_samples,
+        )
+        mapper = generator.flow(G.edges(), edge_labels)
+        assert generator.batch_size == self.batch_size
+        assert mapper.data_size == G.number_of_edges()
+        assert len(mapper.ids) == G.number_of_edges()
+
+    def test_DirectedGraphSAGELinkGenerator_1(self):
+
+        G = example_graph(feature_size=self.n_feat, is_directed=True)
+        data_size = G.number_of_edges()
+        edge_labels = [0] * data_size
+
+        mapper = DirectedGraphSAGELinkGenerator(
+            G,
+            batch_size=self.batch_size,
+            in_samples=self.in_samples,
+            out_samples=self.out_samples,
+        ).flow(G.edges(), edge_labels)
+
+        assert len(mapper) == 2
+
+        for batch in range(len(mapper)):
+            nf, nl = mapper[batch]
+
+            assert len(nf) == 2 ** (len(self.in_samples) + 2) - 2
+
+            for ii in range(len(nf) // 2):
+                assert nf[2 * ii].shape == nf[2 * ii + 1].shape
+
+            # head nodes
+            for ii in range(2):
+                assert nf[ii].shape == (min(self.batch_size, data_size), 1, self.n_feat)
+
+            # first hop in samples
+            for ii in range(2, 4):
```

Comment:
Maybe we can combine all these individual for-loops that look mostly the same:
```python
ins, outs = self.in_samples, self.out_samples
dims = [
    1,
    ins[0], 
    outs[0], 
    ins[0] * outs[1], 
    ins[0] * outs[1], 
    outs[0] * ins[1], 
    outs[0] * outs[1]
]
for i, dim in zip(range(0, 14, 2), dims):
    assert nf[i].shape == nf[i+1].shape == min(self.batch_size, data_size), dim, self.n_feat
```
or something..?

and similarly for `test_DirectedGraphSAGELinkGenerator_unsupervisedSampler_sample_generation`

Reply:
nice!